### PR TITLE
Support admin password specification

### PR DESCRIPTION
**Pull request by abellotti**
- Eliminating the default admin password and relying on the users to specify their default via the environment variable QPC_SERVER_PASSWORD (which also supports the OpenShift/Kubernetes secret declared ones).
  - If not specified, we generate a random password for the user and provide it to them upon creation.
  - When the secret is specified via QPC_SERVER_PASSWORD (either coming from OpenShift or from bash), we also allow the users to update the admin password
 
During server initialization:

```
$ make server-init QPC_SERVER_PASSWORD="SuperAdmin1"
```

As well as any point in time:

```
$ make set-superuser QPC_SERVER_PASSWORD="NewAdminPassword2"
```

When initially creating the superuser and no QPC_SERVER_PASSWORD is specified, a random password is generated:

```
$ make server-init
/Users/abellotti/Library/Caches/pypoetry/virtualenvs/quipucords-WveDg5-x-py3.11/bin/python quipucords/manage.py migrate --settings quipucords.settings -v 3
Operations to perform:
  Apply all migrations: admin, api, auth, authtoken, contenttypes, sessions
Running pre-migrate handlers for application admin
Running pre-migrate handlers for application auth
...
cat ./deploy/setup_user.py | /Users/abellotti/Library/Caches/pypoetry/virtualenvs/quipucords-WveDg5-x-py3.11/bin/python quipucords/manage.py shell --settings quipucords.settings -v 3
Created user admin with random password W2QGYfUUm6
```

With the last line above showing the newly generated password.

Draft for now, needs:
- [x] more testing
- [x] testing from OpenShift instrumenting definiting QPC_SERVER_PASSWORD from a Secret.
- [x] testing/updating camayoc to make sure a QPC_SERVER_PASSWORD is specified and appropriately updated in the camayoc config.yml
- [x] tests added
- [x] verify with enhanced deployer https://github.com/quipucords/quipucords-deployer/pull/22

This PR resolves: https://issues.redhat.com/browse/DISCOVERY-409

A separate issue and PR(s) would be needed for supporting the user password updates via the API (and as a follow-up via the Qpc CLI and finally the quipucords-ui).

### CODE REVIEW

**Comment by abellotti** [comment]
Testing with OpenShift ran as expected ...

With initial deployment as it with current quipucords-deployer (no server password secret):
Screenshot 2023-10-18 at 5 21 25 PM

Adding the server password to the discovery-secrets:

Screenshot 2023-10-18 at 5 39 10 PM

Defining the QPC_SERVER_USERNAME  (admin) and related QPC_SERVER_PASSWORD (picking it up from the secret) in the discovery-server deployment:


Screenshot 2023-10-18 at 5 39 49 PM

Deleting the discovery-server pod so it gets re-deployed:

Screenshot 2023-10-18 at 5 41 13 PM

Screenshot 2023-10-18 at 5 41 21 PM

You can see in the discovery server log that the admin password is getting updated because of the new QPC_SERVER_PASSWORD environment variable:


Screenshot 2023-10-18 at 5 42 31 PM

Finally, logging in via the new password that was specified in the secret:

Screenshot 2023-10-18 at 5 43 06 PM

File: deploy/rejected_common_passwords.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+qpcpassw0rd
```
Comment (mirekdlugosz) [inline deploy/rejected_common_passwords.txt:1]: There should be `qpcpassw0rd` **and** `dscpassw0rd`. The latter is what downstream docs use.

But can we extract rejection logic to another PR, that will be merged few days after this one? This will give us some time to update Jenkins to use new password. Right now we would need to prepare PRs in few repos and merge them all at the same time.

Comment (abellotti) [reply deploy/rejected_common_passwords.txt:1]: 👍 I can keep this file empty (and skip the tests) for now until we're ready from the Jenkins side.

Comment (abellotti) [reply deploy/rejected_common_passwords.txt:1]: np, commit https://github.com/quipucords/quipucords/pull/2502/commits/9ccf78871f291d3117d302b60cad9fd6ae92c8c4 adds dscpassw0rd.

File: quipucords/tests/quipucords/test_user.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    def test_reject_qpc_common_words_qpc(self):
+        """Test that we reject quipucords common words."""
+        with pytest.raises(ValidationError, match="This password is too common"):
+            password_validation.validate_password("qpcpassw0rd")
+
+    def test_reject_qpc_common_words_dsc(self):
+        """Test that we reject quipucords common words."""
+        with pytest.raises(ValidationError, match="This password is too common"):
+            password_validation.validate_password("dscpassw0rd")
+
```
Comment (bruno-fs) [inline quipucords/tests/quipucords/test_user.py:50]: this part can be DRYfied with parametrization =D

```suggestion
@pytest.mark.parametrize("password", ["hello", "qpcpassw0rd", "dscpassw0rd"])
    def test_reject_common_words(self, password):
        """Test that we reject common words."""
        with pytest.raises(ValidationError, match="This password is too common"):
            password_validation.validate_password(password)

```

**Review by bruno-fs — APPROVED** [review]
LGTM

**Comment by mirekdlugosz** [comment]
Two comments from testing:

1. If you try to use common password in container, the logs will mention "Invalid server password specified: - This password is too common.", but Discovery will start normal otherwise. I know it should be possible to recover from this, but would it make sense to fail loudly at this point? Message is easy to miss, container that refused to start is not.

2. There's a lot of fluff printed **after** the password (see below). Could we change it, so password is easy to find right after I start a container?

```
Adding permission 'Permission object (120)'
cat ./deploy/setup_user.py | /opt/venv/bin/python quipucords/manage.py shell --settings quipucords.settings -v 3
Created user admin with random password 5CBZYHQeXa
make: Leaving directory '/app'
[2023-10-26 10:49:50 +0000] [24] [INFO] Starting gunicorn 20.1.0
[2023-10-26 10:49:50 +0000] [24] [INFO] Listening at: https://0.0.0.0:443 (24)
[2023-10-26 10:49:50 +0000] [24] [INFO] Using worker: sync
[2023-10-26 10:49:50 +0000] [24] [INFO] Server is ready. Spawning workers
[2023-10-26 10:49:50 +0000] [25] [INFO] Booting worker with pid: 25
[2023-10-26 10:49:50 +0000] [25] [INFO] Worker spawned (pid: 25)
INFO 2023-10-26 10:49:50,514 environment 25 140276778108736 Platform:
INFO 2023-10-26 10:49:50,515 environment 25 140276778108736 system - Linux 
INFO 2023-10-26 10:49:50,515 environment 25 140276778108736 node - discovery-latest-pod 
INFO 2023-10-26 10:49:50,515 environment 25 140276778108736 release - 6.5.7-200.fc38.x86_64 
INFO 2023-10-26 10:49:50,515 environment 25 140276778108736 version - #1 SMP PREEMPT_DYNAMIC Wed Oct 11 04:07:58 UTC 2023 
INFO 2023-10-26 10:49:50,515 environment 25 140276778108736 machine - x86_64 
INFO 2023-10-26 10:49:50,515 environment 25 140276778108736 processor - x86_64 
INFO 2023-10-26 10:49:50,515 environment 25 140276778108736 Python: 3.11.2 (main, Sep 12 2023, 00:00:00) [GCC 11.3.1 20221121 (Red Hat 11.3.1-4)]
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 Modules: _cffi_backend - 1.16.0, _csv - 1.0, _ctypes - 1.1.0, _curses - b'2.2', _decimal - 1.70, amqp - 5.1.1, amqp.connection - 5.1.1, ansible - 2.15.5, ansible.constants - 2.15.5, ansible.module_utils.six - 1.16.0, ansible.release - 2.15.5, argparse - 1.1, asgiref - 3.7.2, bcrypt - 4.0.1, bcrypt.__about__ - 4.0.1, billiard - 4.1.0, celery - 5.3.4, certifi - 2023.07.22, cffi - 1.16.0, charset_normalizer - 3.3.0, charset_normalizer.version - 3.3.0, click - 8.1.7, cryptography - 41.0.4, cryptography.__about__ - 41.0.4, csv - 1.0, ctypes - 1.1.0, dateutil - 2.8.2, decimal - 1.70, django - 4.2.6, django_filters - 23.3, environ - 0.10.0, gunicorn - 20.1.0, gunicorn.arbiter - 20.1.0, gunicorn.config - 20.1.0, http.server - 0.6, idna - 3.4, idna.idnadata - 15.0.0, idna.package_data - 3.4, ipaddress - 1.0, jinja2 - 3.1.2, json - 2.0.9, kombu - 5.3.2, logging - 0.5.1.2, markupsafe - 2.1.3, packaging - 23.2, pkg_resources._vendor.more_itertools - 9.1.0, pkg_resources._vendor.packaging - 23.1, pkg_resources._vendor.platformdirs - 2.6.2, pkg_resources._vendor.platformdirs.version - 2.6.2, pkg_resources.extern.more_itertools - 9.1.0, pkg_resources.extern.packaging - 23.1, pkg_resources.extern.platformdirs - 2.6.2, platform - 1.0.8, psycopg2 - 2.9.9 (dt dec pq3 ext lo64), psycopg2._psycopg - 2.9.9 (dt dec pq3 ext lo64), pydantic - 1.10.13, pytz - 2023.3.post1, re - 2.2.1, requests - 2.31.0, requests.__version__ - 2.31.0, requests.packages.idna - 3.4, requests.packages.idna.idnadata - 15.0.0, requests.packages.idna.package_data - 3.4, requests.packages.urllib3 - 1.26.17, requests.packages.urllib3._version - 1.26.17, requests.packages.urllib3.connection - 1.26.17, requests.packages.urllib3.packages.six - 1.16.0, requests.packages.urllib3.util.ssl_match_hostname - 3.5.0.1, requests.utils - 2.31.0, rest_framework - 3.14.0, six - 1.16.0, socketserver - 0.4, sqlparse - 0.4.4, urllib.request - 3.11, urllib3 - 1.26.17, urllib3._version - 1.26.17, urllib3.connection - 1.26.17, urllib3.packages.six - 1.16.0, urllib3.util.ssl_match_hostname - 3.5.0.1, vine - 5.0.0, wsgiref.simple_server - 0.2, yaml - 6.0.1, zlib - 1.0
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 -------------------- BEGIN ENVIRONMENT VARIABLES --------------------
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 QPC_DBMS_PASSWORD - ********
HOSTNAME - discovery-latest-pod
QPC_SERVER_TIMEOUT - 120
SSH_AUTH_SOCK - /tmp/ssh-XXXXXXuJ4jwX/agent.2
SSH_AGENT_PID - 3
PWD - /app
DJANGO_LOG_FILE - /var/log/app.log
QPC_LOG_DIRECTORY - /var/log
PRODUCTION - True
QPC_DBMS_USER - dsc
container - oci
HOME - /root
LANG - C.UTF-8
DJANGO_LOG_LEVEL - INFO
DJANGO_LOG_FORMATTER - verbose
PYTHONPATH - /app/quipucords
QUIPUCORDS_LOG_LEVEL - INFO
DJANGO_DEBUG - False
SHLVL - 1
QPC_SERVER_USERNAME - admin
NETWORK_INSPECT_JOB_TIMEOUT - 600
DJANGO_SECRET_PATH - /var/data/secret.txt
NETWORK_CONNECT_JOB_TIMEOUT - 60
QPC_DBMS_HOST - localhost
DJANGO_DB_PATH - /var/data/
QPC_DBMS_DATABASE - dsc-db
QPC_SERVER_USER_EMAIL - admin@example.com
LC_ALL - C.UTF-8
PATH - /opt/venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
DJANGO_LOG_HANDLERS - console,file
_ - /opt/venv/bin/gunicorn
SERVER_SOFTWARE - gunicorn/20.1.0
DJANGO_SETTINGS_MODULE - quipucords.settings
ANSIBLE_HOST_KEY_CHECKING - False
ANSIBLE_NOCOLOR - False
TZ - UTC
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 -------------------- END ENVIRONMENT VARIABLES --------------------
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 QPC_DBMS set to 'postgres'
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 QPC_DBMS_HOST set to 'localhost'
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 QPC_DBMS_PORT set to 5432
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 QPC_DBMS_DATABASE set to 'dsc-db'
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 QPC_DBMS_USER set to 'dsc'
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 REDIS_USERNAME set to ''
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 REDIS_HOST set to 'localhost'
INFO 2023-10-26 10:49:50,517 environment 25 140276778108736 REDIS_PORT set to 6379
INFO 2023-10-26 10:49:50,519 environment 25 140276778108736 Server version: 1.4.0+9d922ef3bd20b9d1ab085deb38cd538bd484d5e1
INFO 2023-10-26 10:49:50,520 environment 25 140276778108736 Commit: 9d922ef3bd20b9d1ab085deb38cd538bd484d5e1
INFO 2023-10-26 10:49:50,525 model 25 140276778108736 Server identification not found.  Initializing server identifier to d42507b2-89af-4257-9f55-36b31ce1b777.
INFO 2023-10-26 10:49:50,541 environment 25 140276778108736 Server ID: d42507b2-89af-4257-9f55-36b31ce1b777
INFO 2023-10-26 10:49:50,688 manager 25 140276778108736 SCAN JOB MANAGER: Scan manager instance created.
INFO 2023-10-26 10:49:50,688 manager 25 140276680123968 SCAN JOB MANAGER: Searching for incomplete scans
INFO 2023-10-26 10:49:50,693 manager 25 140276680123968 SCAN JOB MANAGER: No running or pending scan jobs to start
INFO 2023-10-26 10:49:50,693 manager 25 140276680123968 SCAN JOB MANAGER: Started run loop.
INFO 2023-10-26 10:49:50,693 manager 25 140276680123968 SCAN JOB MANAGER: No scan job currently running.  Scan queue length is 0. Queued jobs: []
```

**Comment by mirekdlugosz** [comment]
One more thing:

```
$ podman exec -ti dsc-latest-backend make server-set-superuser QPC_SERVER_PASSWORD="MySuperSecretPassword"
cat ./deploy/setup_user.py | /opt/venv/bin/python quipucords/manage.py shell --settings quipucords.settings -v 3
Created user admin
```

This message is misleading - user admin was created back when container was created, now only password was changed. This is minor thing, but would it be easy enough to differentiate these two cases and provide better messages?

**Comment by infinitewarp** [comment]
> There's a lot of fluff printed after the password (see below). Could we change it, so password is easy to find right after I start a container?

If you want to omit the environment variables dump, I believe setting `QPC_LOG_ALL_ENV_VARS_AT_STARTUP=False` will do that. Is that the part you consider "fluff", or is it something else, @mirekdlugosz ? (I don't disagree on the "fluff" comment; I just mean to clarify which in this case.)

**Comment by mirekdlugosz** [comment]
@infinitewarp environment variables, but not only. Basically everything after "Created user admin with random password" line. Ideally, I would start a container, look into logs and I should see password right there, without scrolling up.

**Comment by abellotti** [comment]
Regarding `Created user admin` comment above. As discussed offline, the initial server run was with a common password, thus the admin user was never created. When the server-set-superuser command ran, it was indeed a new admin user getting created, otherwise the `User admin already exists, updated password` message would have been seen.

**Comment by abellotti** [comment]
@mirekdlugosz re: the first comment, here's the commit https://github.com/quipucords/quipucords/pull/2502/commits/8df7a8512e6313e2a98c10df4baaed39556ea56b. Essentially we shouldn't start the server if either DB migration or superuser successfully created/updated.
